### ISSUE TITLE
CUB histogram API signatures are misleading

### ISSUE THREAD

**Issue by bernhardmgruber** [NVIDIA/cccl#1765]
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Is this for new documentation, or an update to existing docs?

Update

### Describe the incorrect/future/missing documentation

Many CUB histogram APIs have signatures like this:
```c++
  template <int NUM_CHANNELS,
            int NUM_ACTIVE_CHANNELS,
            ...>
  ... static cudaError_t MultiHistogramEven(
    ...,
    CounterT* d_histogram[NUM_ACTIVE_CHANNELS],
    int num_levels[NUM_ACTIVE_CHANNELS],
    LevelT lower_level[NUM_ACTIVE_CHANNELS],
    LevelT upper_level[NUM_ACTIVE_CHANNELS],
   ...)
```
(See: https://github.com/NVIDIA/cccl/blob/main/cub/cub/device/device_histogram.cuh#L597-L613)

Parameters like `CounterT* d_histogram[NUM_ACTIVE_CHANNELS]` suggest that an array of `NUM_ACTIVE_CHANNELS` values is expected. However, syntactically, the parameter type is just `CounterT**` since the array extent is ignored. The API can in fact be called with arrays of more or less than `NUM_ACTIVE_CHANNELS` values, or with plain pointers. This may be surprising to some users or wrongly encourage them to setup parameters as native arrays, while they could pass arguments of a wider range of types.

A way to enforce the array extent, is to take the array by reference: `CounterT* (&d_histogram)[NUM_ACTIVE_CHANNELS]`. However, this would break all existing code depending on the current array to pointer decay. I therefore suggest to use plain pointers in the API, which better expresses the current API surface:
```c++
  template <int NUM_CHANNELS,
            int NUM_ACTIVE_CHANNELS,
            ...>
  ... static cudaError_t MultiHistogramEven(
    ...,
    CounterT** d_histogram,
    int* num_levels,
    LevelT* lower_level,
    LevelT* upper_level,
   ...)
```

### If this is a correction, please provide a link to the incorrect documentation. If this is a new documentation request, please link to where you have looked.

N/A

**Comment by bernhardmgruber** [comment]
The situation is even more complex. A parameter like `CounterT* d_histogram[NUM_ACTIVE_CHANNELS]`, which is actually a `CounterT** d_histogram`, is a pointer to memory visible to the API caller, while the inner pointers are to device memory. So when `MultiHistogramEven` is called from the 

* host, `CounterT**` is a pointer to host memory containing device pointers. 
* device, `CounterT**` is a pointer to device memory containing device pointers.

The entire problem could be side-stepped if `MultiHistogramEven` would take a `cuda::std::array<T, NUM_ACTIVE_CHANNELS>`. `NUM_ACTIVE_CHANNELS` is typically a small number like 4, 3 or 1.

**Comment by bernhardmgruber** [comment]
@gevtushenko and I discussed this issue and concluded that adding overloads taking `cuda::std::array` would be beneficial.

**Comment by Revaj** [comment]
Hello, is this issue still available? I would like to work with it :)

**Comment by bernhardmgruber** [comment]
Hi! Thanks for your interest in working on this issue! In a discussion with @miscco a while back, we considered `::cuda::std::span` as a good candidate for the public API. However, this will require us to wait for the final drop of C++11 support, since our span implementation requires C++14. Until then, I think we should wait with proposing a concrete PR. You are free to suggest an API design though!

**Comment by ijpq** [comment]
The reason why don't support C++11 buildchain alongwith C++14, is from the maintenance cost perspective, right?

> Hi! Thanks for your interest in working on this issue! In a discussion with [@miscco](https://github.com/miscco) a while back, we considered `::cuda::std::span` as a good candidate for the public API. However, this will require us to wait for the final drop of C++11 support, since our span implementation requires C++14. Until then, I think we should wait with proposing a concrete PR. You are free to suggest an API design though!

**Comment by bernhardmgruber** [comment]
> The reason why don't support C++11 buildchain along with C++14, is from the maintenance cost perspective, right?

Yes. Also developer happiness, since we can only use the oldest C++ dialect in our implementations. We are very happy to have C++17 since the beginning of the year.
